From e558b5e0661740f56a5b8772fd8f99f7b58b7eb9 Mon Sep 17 00:00:00 2001
From: Calvin Walton <calvin.walton@kepstin.ca>
Date: Sun, 29 Mar 2020 14:10:35 -0400
Subject: [PATCH] webkit: bump to 2.28.2

I've taken the opportunity to go through the options we have in the
comments and enabled, and I've removed all of the options that are
either now already enabled by default in the GTK port, or which we have
no business modifying ourselves. I've added some notes about the
"EXPERIMENTAL" options that we might want to look at enabling at some
point.

I've unconditionally applied the alpine patch that changes linux
detection to use the __linux__ macro instead of __GLIBC__. Rather than
apply alpine's patch which reduces stack usage, I've added LDFLAGS which
should configure musl to use 2MB stacks, which matches the glibc
default. See discussion in https://bugs.webkit.org/show_bug.cgi?id=187485

We should really switch to the USE_WPE_RENDERER build... doing so
requires packagine libwpe and WPEBackend-fdo which should be
straightforwards.
---
 .../files/webkit-2.28.2-musl-fixes.patch      | 72 +++++++++++++++++++
 ...28.0-undefined-EGL_WAYLAND_BUFFER_WL.patch | 22 ++++++
 .../webkit-Fix-build-with-icu-65.1.patch      | 56 ---------------
 ...1-r1.exheres-0 => webkit-2.28.2.exheres-0} | 56 +++++++--------
 packages/net-libs/webkit/webkit.exlib         |  5 ++
 5 files changed, 123 insertions(+), 88 deletions(-)
 create mode 100644 packages/net-libs/webkit/files/webkit-2.28.2-musl-fixes.patch
 create mode 100644 packages/net-libs/webkit/files/webkit-2.28.2-undefined-EGL_WAYLAND_BUFFER_WL.patch
 delete mode 100644 packages/net-libs/webkit/files/webkit-Fix-build-with-icu-65.1.patch
 rename packages/net-libs/webkit/{webkit-2.26.1-r1.exheres-0 => webkit-2.28.2.exheres-0} (81%)

diff --git a/packages/net-libs/webkit/files/webkit-2.28.2-musl-fixes.patch b/packages/net-libs/webkit/files/webkit-2.28.2-musl-fixes.patch
new file mode 100644
index 000000000..3f687e170
--- /dev/null
+++ b/packages/net-libs/webkit/files/webkit-2.28.2-musl-fixes.patch
@@ -0,0 +1,72 @@
+Source: https://git.alpinelinux.org/aports/tree/community/webkit2gtk?id=e284ee4af8b2ebe9699a7125205be34db120bfc3
+Reason: Fix compilation with musl libc
+Upstream: not yet. musl tracking bug: https://bugs.webkit.org/show_bug.cgi?id=187485
+
+--- a/Source/JavaScriptCore/runtime/MachineContext.h
++++ b/Source/JavaScriptCore/runtime/MachineContext.h
+@@ -188,7 +188,7 @@ static inline void*& stackPointerImpl(mcontext_t& machineContext)
+ #error Unknown Architecture
+ #endif
+ 
+-#elif OS(FUCHSIA) || defined(__GLIBC__) || defined(__BIONIC__)
++#elif OS(FUCHSIA) || defined(__linux__)
+ 
+ #if CPU(X86)
+     return reinterpret_cast<void*&>((uintptr_t&) machineContext.gregs[REG_ESP]);
+@@ -335,7 +335,7 @@ static inline void*& framePointerImpl(mcontext_t& machineContext)
+ #error Unknown Architecture
+ #endif
+ 
+-#elif OS(FUCHSIA) || defined(__GLIBC__) || defined(__BIONIC__)
++#elif OS(FUCHSIA) || defined(__linux__)
+ 
+ // The following sequence depends on glibc's sys/ucontext.h.
+ #if CPU(X86)
+@@ -482,7 +482,7 @@ static inline void*& instructionPointerImpl(mcontext_t& machineContext)
+ #error Unknown Architecture
+ #endif
+ 
+-#elif OS(FUCHSIA) || defined(__GLIBC__) || defined(__BIONIC__)
++#elif OS(FUCHSIA) || defined(__linux__)
+ 
+ // The following sequence depends on glibc's sys/ucontext.h.
+ #if CPU(X86)
+@@ -639,7 +639,7 @@ inline void*& argumentPointer<1>(mcontext_t& machineContext)
+ #error Unknown Architecture
+ #endif
+ 
+-#elif OS(FUCHSIA) || defined(__GLIBC__) || defined(__BIONIC__)
++#elif OS(FUCHSIA) || defined(__linux__)
+ 
+ // The following sequence depends on glibc's sys/ucontext.h.
+ #if CPU(X86)
+@@ -756,7 +756,7 @@ inline void*& llintInstructionPointer(mcontext_t& machineContext)
+ #error Unknown Architecture
+ #endif
+ 
+-#elif OS(FUCHSIA) || defined(__GLIBC__) || defined(__BIONIC__)
++#elif OS(FUCHSIA) || defined(__linux__)
+ 
+ // The following sequence depends on glibc's sys/ucontext.h.
+ #if CPU(X86)
+ 
+--- a/Source/WebCore/xml/XPathGrammar.cpp
++++ b/Source/WebCore/xml/XPathGrammar.cpp
+@@ -966,7 +966,7 @@ int yydebug;
+ #if YYERROR_VERBOSE
+ 
+ # ifndef yystrlen
+-#  if defined __GLIBC__ && defined _STRING_H
++#  if defined __linux__ && defined _STRING_H
+ #   define yystrlen strlen
+ #  else
+ /* Return the length of YYSTR.  */
+@@ -989,7 +989,7 @@ yystrlen (yystr)
+ # endif
+ 
+ # ifndef yystpcpy
+-#  if defined __GLIBC__ && defined _STRING_H && defined _GNU_SOURCE
++#  if defined __linux__ && defined _STRING_H && defined _GNU_SOURCE
+ #   define yystpcpy stpcpy
+ #  else
+ /* Copy YYSRC to YYDEST, returning the address of the terminating '\0' in
diff --git a/packages/net-libs/webkit/files/webkit-2.28.2-undefined-EGL_WAYLAND_BUFFER_WL.patch b/packages/net-libs/webkit/files/webkit-2.28.2-undefined-EGL_WAYLAND_BUFFER_WL.patch
new file mode 100644
index 000000000..84dd04a5f
--- /dev/null
+++ b/packages/net-libs/webkit/files/webkit-2.28.2-undefined-EGL_WAYLAND_BUFFER_WL.patch
@@ -0,0 +1,22 @@
+Upstream: https://trac.webkit.org/changeset/253553/webkit
+[GTK] Build with USE_WPE_RENDERER=No fails with undefined EGL_WAYLAND_BUFFER_WL
+
+https://bugs.webkit.org/show_bug.cgi?id=205250
+
+Reviewed by Carlos Garcia Campos.
+
+Define the enum if not present.
+
+No new tests, just a build fix.
+
+--- a/Source/WebKit/UIProcess/gtk/WaylandCompositor.cpp
++++ b/Source/WebKit/UIProcess/gtk/WaylandCompositor.cpp
+@@ -59,4 +59,8 @@
+ #endif
+ 
++#if !defined(EGL_WAYLAND_BUFFER_WL)
++#define EGL_WAYLAND_BUFFER_WL 0x31D5
++#endif
++
+ #if !defined(PFNEGLQUERYWAYLANDBUFFERWL)
+ typedef EGLBoolean (*PFNEGLQUERYWAYLANDBUFFERWL) (EGLDisplay, struct wl_resource*, EGLint attribute, EGLint* value);
diff --git a/packages/net-libs/webkit/files/webkit-Fix-build-with-icu-65.1.patch b/packages/net-libs/webkit/files/webkit-Fix-build-with-icu-65.1.patch
deleted file mode 100644
index 4728dadca..000000000
--- a/packages/net-libs/webkit/files/webkit-Fix-build-with-icu-65.1.patch
+++ /dev/null
@@ -1,56 +0,0 @@
-Upstream: under review, https://bugs.webkit.org/show_bug.cgi?id=202600
-
-From 9b60e834454dc93f46f05b1cfdc0aad0c6b7de97 Mon Sep 17 00:00:00 2001
-From: Heiko Becker <heirecka@exherbo.org>
-Date: Fri, 4 Oct 2019 22:17:11 +0200
-Subject: [PATCH] Add missing semicolons to fix build with icu 65.1
-
----
- Source/WTF/wtf/URLHelpers.cpp   |  2 +-
- Source/WebCore/dom/Document.cpp |  6 +++---
- 2 files changed, 4 insertions(+), 4 deletions(-)
-
-diff --git a/Source/WTF/wtf/URLHelpers.cpp b/Source/WTF/wtf/URLHelpers.cpp
-index 18e7f13cd61..c584f1a0cb7 100644
---- a/Source/WTF/wtf/URLHelpers.cpp
-+++ b/Source/WTF/wtf/URLHelpers.cpp
-@@ -301,7 +301,7 @@ static bool allCharactersInIDNScriptWhiteList(const UChar* buffer, int32_t lengt
-     Optional<UChar32> previousCodePoint;
-     while (i < length) {
-         UChar32 c;
--        U16_NEXT(buffer, i, length, c)
-+        U16_NEXT(buffer, i, length, c);
-         UErrorCode error = U_ZERO_ERROR;
-         UScriptCode script = uscript_getScript(c, &error);
-         if (error != U_ZERO_ERROR) {
-diff --git a/Source/WebCore/dom/Document.cpp b/Source/WebCore/dom/Document.cpp
-index 0fd0fa090bf..01f76850f3d 100644
---- a/Source/WebCore/dom/Document.cpp
-+++ b/Source/WebCore/dom/Document.cpp
-@@ -4954,12 +4954,12 @@ static bool isValidNameNonASCII(const UChar* characters, unsigned length)
-     unsigned i = 0;
- 
-     UChar32 c;
--    U16_NEXT(characters, i, length, c)
-+    U16_NEXT(characters, i, length, c);
-     if (!isValidNameStart(c))
-         return false;
- 
-     while (i < length) {
--        U16_NEXT(characters, i, length, c)
-+        U16_NEXT(characters, i, length, c);
-         if (!isValidNamePart(c))
-             return false;
-     }
-@@ -5019,7 +5019,7 @@ ExceptionOr<std::pair<AtomString, AtomString>> Document::parseQualifiedName(cons
- 
-     for (unsigned i = 0; i < length; ) {
-         UChar32 c;
--        U16_NEXT(qualifiedName, i, length, c)
-+        U16_NEXT(qualifiedName, i, length, c);
-         if (c == ':') {
-             if (sawColon)
-                 return Exception { InvalidCharacterError };
--- 
-2.23.0
-
diff --git a/packages/net-libs/webkit/webkit-2.26.1-r1.exheres-0 b/packages/net-libs/webkit/webkit-2.28.2.exheres-0
similarity index 81%
rename from packages/net-libs/webkit/webkit-2.26.1-r1.exheres-0
rename to packages/net-libs/webkit/webkit-2.28.2.exheres-0
index e56c41470..18e346f51 100644
--- a/packages/net-libs/webkit/webkit-2.26.1-r1.exheres-0
+++ b/packages/net-libs/webkit/webkit-2.28.2.exheres-0
@@ -2,7 +2,7 @@
 # Copyright 2017 Rasmus Thomsen <Rasmus.thomsen@protonmail.com>
 # Distributed under the terms of the GNU General Purpose License v2
 
-require cmake [ api=2 ] webkit [ required_gcc='6.0' ]
+require cmake [ api=2 ninja=true ] webkit [ required_gcc='7.3' ]
 require utf8-locale
 
 SLOT="4.0"
@@ -42,7 +42,7 @@ DEPENDENCIES="
         dev-libs/atk[>=2.16.0]
         dev-libs/glib:2[>=2.44.0] [[ note = [ 2.44 for glib networking ] ]]
         dev-libs/icu:=
-        dev-libs/libgcrypt[>=1.6.0]
+        dev-libs/libgcrypt[>=1.7.0]
         dev-libs/libsecret:1
         dev-libs/libxml2:2.0[>=2.8.0]
         dev-libs/libxslt[>=1.1.7]
@@ -62,14 +62,14 @@ DEPENDENCIES="
         x11-dri/mesa[X?][wayland?]
         x11-libs/cairo[>=1.16][X?]
         x11-libs/gtk+:3[>=3.22.0][gobject-introspection?][X?][wayland?]
-        x11-libs/harfbuzz[>=0.9.7]
+        x11-libs/harfbuzz[>=0.9.18]
         x11-libs/pango[>=1.32.0]
         gobject-introspection? ( gnome-desktop/gobject-introspection:1[>=1.32.0] )
         hyphen? ( office-libs/hyphen )
         jpeg2000? ( media-libs/OpenJPEG:2[>=2.2.0] )
         libnotify? ( x11-libs/libnotify )
         opengl? ( media-plugins/gst-plugins-base:1.0[gstreamer_plugins:opengl] )
-        spell? ( app-spell/enchant:0[>=0.22] )
+        spell? ( app-spell/enchant:2 )
         X? (
             x11-libs/libXcomposite
             x11-libs/libXdamage
@@ -87,57 +87,49 @@ DEPENDENCIES="
 "
 
 DEFAULT_SRC_PREPARE_PATCHES=(
-    "${FILES}"/${PN}-Fix-build-with-icu-65.1.patch
+    "${FILES}"/${PN}-2.28.2-musl-fixes.patch
+    "${FILES}"/${PN}-2.28.2-undefined-EGL_WAYLAND_BUFFER_WL.patch
 )
 
 CMAKE_SOURCE="${WORKBASE}/webkitgtk-${PV}/"
 
 # ENABLE_ACCELERATED_2D_CANVAS
 #   conflicts with ENABLE_GLES2
-# ENABLE_DATA_TRANSFER_ITEMS
-# ENABLE_INPUT_TYPE_COLOR
-# ENABLE_INPUT_TYPE_DATE
-# ENABLE_INPUT_TYPE_DATETIMELOCAL
-# ENABLE_INPUT_TYPE_MONTH
-# ENABLE_INPUT_TYPE_TIME
-# ENABLE_INPUT_TYPE_WEEK
-# ENABLE_MEDIA_STATISTICS
-# ENABLE_SCRIPTED_SPEECH
-# ENABLE_TEXT_AUTOSIZING
 # ENABLE_WEBASSEMBLY
-#   disabled by default upstream
+#   Automatically enabled when ENABLE_FTL_JIT:BOOL=ON
 # USE_SYSTEM_MALLOC
 #   disabled by default upstream
 #   NOTE(compnerd) DO NOT PROVIDE AN OPTION FOR THIS (bmalloc is only disabled on android)
-# ENABLE_CSS_COMPOSITING
+# ENABLE_TEXT_AUTOSIZING
 # ENABLE_DEVICE_ORIENTATION
 # ENABLE_ORIENTATION_EVENTS
-#   only enabled on iOS builds
-# ENABLE_MOUSE_CURSOR_SCALE
+#   only enabled on iOS builds, maybe useful for mobile device targets?
 # ENABLE_TOUCH_SLIDER
 #   only enabled on enlightenment builds
 #   TODO(compnerd) determine what these do
 # ENABLE_FTPDIR
-# ENABLE_INDEXED_DATABASE_IN_WORKERS
-#   disabled by default in GTK+ port
 # ENABLE_WEBGPU
 #   New alternative to WebGL, experimental
 # ENABLE_WEBGL2
-#   fails to build as of 2.24.1
-# DENABLE_MEDIA_STREAM:BOOL=OFF
-# DENABLE_WEB_RTC:BOOL=OFF
-#   Requires unpackaged OpenWebRTC
+#   fails to build as of 2.28.2
+#
+# ENABLE_ENCRYPTED_MEDIA
+# ENABLE_MEDIA_STREAM
+# ENABLE_OFFSCREEN_CANVAS
+# ENABLE_WEB_RTC
+# ENABLE_CSS_PAINTING_API
+# ENABLE_CSS_TYPED_OM
+# ENABLE_POINTER_EVENTS
+#   Under ENABLE_EXPERIMENTAL_FEATURES in GTK port.
+#   WEB_RTC & MEDIA_STREAM Uses gstreamer & bundled libwebrtc
+#
+# ENABLE_REMOTE_INSPECTOR PRIVATE
+# ENABLE_RESOURCE_USAGE
+#   Under ENABLE_EXPERIMENTAL_FEATURES in all ports.
 
 CMAKE_SRC_CONFIGURE_PARAMS=(
-    -DENABLE_BUBBLEWRAP_SANDBOX:BOOL=ON
-    -DENABLE_VIDEO:BOOL=ON
-    -DENABLE_WEB_AUDIO:BOOL=ON
-    -DENABLE_WEB_CRYPTO:BOOL=ON
-    -DENABLE_WEBDRIVER:BOOL=ON
-    -DENABLE_XSLT:BOOL=ON
     -DPORT=GTK
     -DUSE_LD_GOLD:BOOL=OFF
-    -DUSE_LIBSECRET:BOOL=ON
 
     # Needs unpackaged libwpe
     -DUSE_WPE_RENDERER:BOOL=OFF
diff --git a/packages/net-libs/webkit/webkit.exlib b/packages/net-libs/webkit/webkit.exlib
index 61a4b274a..845b8936a 100644
--- a/packages/net-libs/webkit/webkit.exlib
+++ b/packages/net-libs/webkit/webkit.exlib
@@ -43,6 +43,11 @@ webkit_src_prepare() {
     append-ldflags -Wl,--no-keep-memory
     [[ $(eclectic ld show) == bfd ]] && append-ldflags -Wl,--reduce-memory-overheads
 
+    # Webkit doesn't work with musl's default (small) thread stack sizes.
+    # https://bugs.webkit.org/show_bug.cgi?id=187485
+    # Set the stack size on musl to match glibc's x86 default 2MB
+    [[ $(exhost --target) == *-musl* ]] && append-ldflags -Wl,-z,stack-size=2097152
+
     # because we must stuff everything into an exlib
     if ever at_least 2.20.2 ; then
         cmake_src_prepare
-- 
2.25.2

